tests/tools/plugins/ssl_hook_test.cc

```diff
@@ -0,0 +1,248 @@
+/** @file
+
+  SSL Preaccept test plugin
+  Implements blind tunneling based on the client IP address
+  The client ip addresses are specified in the plugin's
+  config file as an array of IP addresses or IP address ranges under the
+  key "client-blind-tunnel"
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include <ts/ts.h>
+#include <ts/remap.h>
+#include <getopt.h>
+#include <openssl/ssl.h>
+
+#define PN "ssl_hook_test"
+#define PCP "[" PN " Plugin] "
+
+int
+ReenableSSL(TSCont cont, TSEvent event, void *edata)
+{
+  TSVConn ssl_vc = reinterpret_cast<TSVConn>(TSContDataGet(cont));
+  TSVConnReenable(ssl_vc);
+  TSContDestroy(cont);
+  TSDebug(PN, "Callback reenable ssl_vc=%p", ssl_vc);
+  return TS_SUCCESS;
+}
+
+int
+CB_Pre_Accept(TSCont cont, TSEvent event, void *edata)
+{
+  TSVConn ssl_vc = reinterpret_cast<TSVConn>(edata);
+
+  int count = reinterpret_cast<intptr_t>(TSContDataGet(cont));
+
+  TSDebug(PN, "Pre accept callback %d %p - event is %s", count, ssl_vc, event == TS_EVENT_VCONN_PRE_ACCEPT ? "good" : "bad");
+
+  // All done, reactivate things
+  TSVConnReenable(ssl_vc);
+  return TS_SUCCESS;
+}
+
+int
+CB_Pre_Accept_Delay(TSCont cont, TSEvent event, void *edata)
+{
+  TSVConn ssl_vc = reinterpret_cast<TSVConn>(edata);
+
+  int count = reinterpret_cast<intptr_t>(TSContDataGet(cont));
+
+  TSDebug(PN, "Pre accept delay callback %d %p - event is %s", count, ssl_vc, event == TS_EVENT_VCONN_PRE_ACCEPT ? "good" : "bad");
+
+  TSCont cb = TSContCreate(&ReenableSSL, TSMutexCreate());
+
+  TSContDataSet(cb, ssl_vc);
+
+  // Schedule to reenable in a bit
+  TSContSchedule(cb, 2000, TS_THREAD_POOL_NET);
+
+  return TS_SUCCESS;
+}
+
+int
+CB_SNI(TSCont cont, TSEvent event, void *edata)
```

Comment:
The crash I was seeing in ubuntu was due to SSL_get_servername((SSL*)ssl_vc, ). The following code worked for me
```
    int

    CB_SNI(TSCont cont, TSEvent event, void *edata)

    {

      TSVConn ssl_vc         = reinterpret_cast<TSVConn>(edata);

    TSSslConnection ssl_conn = TSVConnSSLConnectionGet(ssl_vc);

      auto *ssl                = reinterpret_cast<SSL *>(ssl_conn);

      char const *servername          = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);

     

    if(ssl_vc){

      //const char *servername = SSL_get_servername((SSL *)ssl_vc, TLSEXT_NAMETYPE_host_name);

     TSDebug(PN, "reallyyyyy   SNI callback - SNI is %p",edata);

      int count = reinterpret_cast<intptr_t>(TSContDataGet(cont));

      TSDebug(PN, "SNI callback %d %p - SNI is %s", count, ssl_vc, servername);

      // All done, reactivate things

      TSVConnReenable(ssl_vc);

    }

      return TS_SUCCESS;

    }
```